### ISSUE TITLE
Generated Protos are no longer considered as virtual_imports in Bazel 7

### ISSUE THREAD

**Issue by XuanWang-Amos** [bazelbuild/bazel#21075]
### Description of the bug:

[One of our test](https://github.com/grpc/grpc/blob/714640d71f07d415a14aab2c68e8f37d74c644aa/test/distrib/bazel/python/BUILD#L123) starts failing in Bazel 7, after some investigation, we found out it's because Bazel 7.0.0 removed the generated file check and stopped symlink our generated proto to `_virtual_imports`.

* Bazel 6.4.0 will symlink generated proto to `_virtual_imports` because it's a generated file ([link to code](https://github.com/bazelbuild/bazel/blob/50b61e38fb6eaa08a6c811a313824a59f1c2abbf/src/main/starlark/builtins_bzl/common/proto/proto_library.bzl#L94-L106)):
```
    generate_protos_in_virtual_imports = False
    if ctx.fragments.proto.generated_protos_in_virtual_imports():
        generate_protos_in_virtual_imports = any([not src.is_source for src in srcs])

    if import_prefix != "" or strip_import_prefix != "" or generate_protos_in_virtual_imports:
        # Use virtual source roots
        return _symlink_to_virtual_imports(ctx, srcs, import_prefix, strip_import_prefix)
```

* Bazel 7.0.0 **removed the generated file check** thus it will not symlink generated proto to `_virtual_imports` ([link to code](https://github.com/bazelbuild/bazel/blob/d798ebde6c6394203a87b5f1a6b62ecfc3880991/src/main/starlark/builtins_bzl/common/proto/proto_library.bzl#L113)):
```
    if import_prefix != "" or strip_import_prefix != "":
        # Use virtual source roots
        return _symlink_to_virtual_imports(ctx, srcs, import_prefix, strip_import_prefix)
    else:
        # No virtual source roots
        return "", srcs
```

Is this change of behavior expected? If it is, can someone share some context on why it was changed?

Link to our issue: https://github.com/grpc/grpc/issues/35391

### Which category does this issue belong to?

Rules API

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Build this with Bazel 7: https://github.com/grpc/grpc/blob/714640d71f07d415a14aab2c68e8f37d74c644aa/test/distrib/bazel/python/BUILD#L123

### Which operating system are you running Bazel on?

Not related

### What is the output of `bazel info release`?

Not related

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

Not related

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?

```text
Not related
```

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

No

### Have you found anything relevant by searching the web?

No

**Comment by comius** [comment]
> Is this change of behavior expected? If it is, can someone share some context on why it was changed?

Yes, this change of behaviour is intentional and expected (I just wish I caught it earlier in grpc). `_virtual_imports` are created only when it's strictly necessary. That's when `proto_library` uses `import_prefix` or `strip_import_prefix`.

The reason for the change was, that the behaviour was is some cases incorrect (external repos using sibling repository layout) and creating `_virtual_imports` created a performance penalty.

Ideally proto generation would happen using `proto_common.compile`, where you don't need to specify output directories anymore. A quicker workaround would be to copy `_output_directory` implementation from `proto_common.bzl` in Bazel.

**Comment by XuanWang-Amos** [comment]
Thanks for the reply, do you mind share more details on the workaround?

By `copy _output_directory implementation` you mean we can use `_output_directory` to replace the outputs [here](https://github.com/grpc/grpc/blob/714640d71f07d415a14aab2c68e8f37d74c644aa/bazel/protobuf.bzl#L205)? What about the parameters, I guess we can use `copied_proto` as `root`, what should be our `proto_info` here?

**Comment by XuanWang-Amos** [comment]
@comius We're still trying to figure out the correct fix, any further details you can provide would be extremely helpful.

**Comment by nickdesaulniers** [comment]
@XuanWang-Amos thank you for the detailed investigation! (And @comius for the additional background)

I recently hit this trying to upgrade a codebase from bazel 6 to bazel 7.

We had a few `proto_library`'s where our `#include` preprocessor statements looked like:
```c++
#include "<additional junk we added>/_virtual_imports/<proto_library name>/<proto src>.pb.h"
```
which broke upon the upgrade to bazel 7 (for the reasons listed above).  The fix for us was simply:

```diff
- #include "<additional junk we added>/_virtual_imports/<proto_library name>/<proto src>.pb.h"
+ #include "<proto src>.pb.h"
```
(We can probably get `<additional junk we added>` back via `import_prefix`, but that causes `import` failures in other `proto_library`'s that depend on these; and we don't control the source of these proto's).

Thank you!
